### PR TITLE
docs: design of Nexus

### PR DESCRIPTION
**Pull request by homura**
preview: https://github.com/ckb-js/nexus/blob/pre-design/docs/design.md
prototype: https://www.figma.com/file/6il5bkgZDrfJhCBs2c7er4/Nexus-Wallet

### 2023-02-09

- a new `ckb_purpose` field is introduced
- change rule-based ownership path to `m / ckb_purpose 1' / account'/ index`
- renaming `getUsedLocks` to `getOnChainLocks`
- renaming `getUnusedLocks` to `getOffChainLocks`

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -47,10 +47,34 @@ The P2SH transaction in CKB looks like this
   <img width="480" src="./design/p2sh.png" />
 </p>
 
-To differentiate from the BIP-44 derivation rules, Nexus can use the [BIP-49](https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki) derivation rules to manage these keys for P2SH
+Unlike full ownership, rule-based ownership tends to exist on-chain as script hash, and because of the flexibility of CKB scripts, script hash does not exist as a fixed `OP_HASH160 [20-byte-hash-value] OP_EQUAL`, but appears on-chain dynamically in various scripts with business logic. As a result, it is difficult for wallets to probe rule-based locks that have already been used.
 
 ```
-m / 49' / 309' / 0' / change / index
+`m / ckb_purpose 1' / account' / index
```
Comment (zhangyouxin) [inline docs/design.md:53]: remove the ` before m

File: docs/design.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+  getNetworkName(): Promise<Network>;
+
+  fullOwnership: Ownership;
+  ruleBasedOwnership: Ownership;
+}
+
+interface FullOwnership {
+  getOffChainLocks(options?: GetOffChainLocksOptions): Promise<Script[]>;
+
+  getOnChainLockss(payload?: { cursor?: string }): Promise<Paginate<Script>>;
```
Comment (gpBlockchain) [inline docs/design.md:114]:
```suggestion
  getOnChainLocks(payload?: { cursor?: string }): Promise<Paginate<Script>>;
```

File: docs/design.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+   * Get the connected network
+   *
+   */
+  getNetworkName(): Promise<Network>;
+
+  fullOwnership: Ownership;
+  ruleBasedOwnership: Ownership;
+}
+
+interface FullOwnership {
```
Comment (gpBlockchain) [inline docs/design.md:111]:
```suggestion
interface Ownership {
```

File: docs/design.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+interface WalletEventListener {
+  on(event: 'networkChanged', cb: (payload: { network: Network }) => void): void;
+}
+
+/**
+ * @see https://github.com/nervosnetwork/ckb/blob/develop/rpc/README.md#type-chaininfo
+ */
+type Network = 'ckb' | 'ckb_testnet';
+
+type GetOffChainLocksOptions = {
```
Comment (zhangyouxin) [inline docs/design.md:135]:
```suggestion
export interface GetPaginateItemsPayload {
  cursor?: string;
}

export interface OwnershipFilter {
  change?: 'external' | 'internal'; 
}

export interface GetLiveCellsPayload extends GetPaginateItemsPayload, OwnershipFilter{}
export interface GetOnChainLocksPayload extends OwnershipFilter, GetPaginateItemsPayload {}
export interface GetOffChainLocksPayload extends OwnershipFilter {};
```

File: docs/design.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  fullOwnership: Ownership;
+  ruleBasedOwnership: Ownership;
+}
+
+interface Ownership {
+  getOffChainLocks(options?: GetOffChainLocksOptions): Promise<Script[]>;
+
+  getOnChainLocks(payload?: { cursor?: string }): Promise<Paginate<Script>>;
+
+  getLiveCells(payload?: { cursor?: string }): Promise<Paginate<Cell>>;
```
Comment (zhangyouxin) [inline docs/design.md:116]:
```suggestion
  getLiveCells(payload?: GetLiveCellsPayload): Promise<Paginate<Cell>>;
```

File: docs/design.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  getNetworkName(): Promise<Network>;
+
+  fullOwnership: Ownership;
+  ruleBasedOwnership: Ownership;
+}
+
+interface Ownership {
+  getOffChainLocks(options?: GetOffChainLocksOptions): Promise<Script[]>;
+
+  getOnChainLocks(payload?: { cursor?: string }): Promise<Paginate<Script>>;
```
Comment (zhangyouxin) [inline docs/design.md:114]:
```suggestion
  getOnChainLocks(payload: GetOnChainLocksPayload): Promise<Paginate<Script>>;
```

File: docs/design.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  * Get the current wallet's nickname for display puropose
+  */
+  getNickname(): Promise<string>;
+
+  fullOwnership: Ownership;
+  ruleBasedOwnership: Ownership;
+}
+
+interface Ownership {
+  getOffChainLocks(options?: GetOffChainLocksOptions): Promise<Script[]>;
```
Comment (gpBlockchain) [inline docs/design.md:117]:
```suggestion
  getOffChainLocks(payload?: GetOffChainLocksPayload): Promise<Script[]>;
```

File: docs/design.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+  cursor?: string;
+}
+
+export interface OwnershipFilter {
+  change?: 'external' | 'internal'; 
+}
+
+export interface GetLiveCellsPayload extends OwnershipFilter, GetPaginateItemsPayload{}
+export interface GetOnChainLocksPayload extends OwnershipFilter, GetPaginateItemsPayload {}
+export interface GetOffChainLocksPayload extends OwnershipFilter {};
```
Comment (gpBlockchain) [inline docs/design.md:150]:
```suggestion
export interface GetOffChainLocksPayload extends OwnershipFilter {
```

File: docs/design.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  // will be ignored in rule-based ownership
+  change?: boolean;
```
Comment (gpBlockchain) [inline docs/design.md:152]: Is `change` here a member variable of `GetOffChainLocksPayload`? It seems to be duplicated with the `change` in OwnershipFilter.

Comment (homura) [reply docs/design.md:152]: the `change` is wrong, it should be deleted

**Comment by homura** [comment]
When I was implementing ownership, I found that there is actually not much difference between rule-based ownership and full ownership. 
Initially, I wanted to use rule-based ownership to simulate P2SH in BIP-49. However, there is not much difference between rule-based ownership and full ownership. 
Perhaps we should only provide full ownership, which is sufficient, otherwise it may bring confusion to developers.

**Pull request merged**